youxia-common/src/main/java/io/cloudbindle/youxia/azure/resourceManagerWrapper/AzureResourceManagerClient.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpPatch;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.DefaultHttpClient;
+
+/**
+ * This wraps calls to the REST-based Resource Manager service on Azure.
+ *
+ * @author dyuen
+ */
+public class AzureResourceManagerClient {
+
+    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    private AuthenticationResult getAccessTokenFromUserCredentials() throws Exception {
+        // this user is an "organization user" microsoft accounts do not work
+        // see http://blog.baslijten.com/create-an-organizational-account-to-administrate-azure-when-having-a-microsoft-account/
+        // the password is set by logging in as the organization user (which will prompt you to set the password)
+        HierarchicalINIConfiguration youxiaConfig = ConfigTools.getYouxiaConfig();
+        String username = youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_USERNAME);
+        String password = youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_PASSWORD);
+        // the end part of this is from the URL of the active directory page, seriously
+        String authority = "https://login.windows.net/" + youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_TENANT_ID);
+        // this is listed under the applications tab and is listed as CLIENT ID
+        String clientID = youxiaConfig.getString(ConfigTools.YOUXIA_AZURE_ACTIVE_DIRECTORY_CLIENT_ID);
+
+        AuthenticationContext context;
+        AuthenticationResult result = null;
+        ExecutorService service = null;
+        try {
+            service = Executors.newFixedThreadPool(1);
+            context = new AuthenticationContext(authority, false, service);
+            Future<AuthenticationResult> future = context.acquireToken("https://management.azure.com/", clientID, username, password, null);
+            result = future.get();
+        } finally {
+            service.shutdown();
+        }
+
+        if (result == null) {
+            throw new ServiceUnavailableException("authentication result was null");
+        }
+        return result;
+    }
+
+    private String getEndpoint(String endpoint) throws Exception {
+        String accessToken;
+        HierarchicalINIConfiguration youxiaConfig = ConfigTools.getYouxiaConfig();
+        AuthenticationResult accessTokenFromUserCredentials = getAccessTokenFromUserCredentials();
+        accessToken = accessTokenFromUserCredentials.getAccessToken();
+
+        HttpClient client = new DefaultHttpClient();
+        HttpGet request = new HttpGet("https://management.azure.com/subscriptions/" + youxiaConfig.getString(YOUXIA_AZURE_SUBSCRIPTION_ID)
+                + "/" + endpoint + "?api-version=2015-01-01");
```

Comment:
What happens if the API version changes? Are there actually different API version avilable?